R/map-if-at.R

```diff
@@ -0,0 +1,143 @@
+  sel <- probe(.x, .p)
+
+  out <- list_along(.x)
+  out[sel]  <- map(.x[sel], .f, ...)
+
+  if (is_null(.else)) {
+    out[!sel] <- .x[!sel]
+  } else {
+    out[!sel]  <- map(.x[!sel], .else, ...)
+  }
+
+  set_names(out, names(.x))
+}
+#' @rdname map_if
+#' @param .at A character vector of names, positive numeric vector of
+#'   positions to include, or a negative numeric vector of positions to
+#'   exlude. Only those elements corresponding to `.at` will be modified.
+#'
+#'   `r lifecycle::badge("deprecated")`: if the tidyselect package is
+#'   installed, you can use `vars()` and tidyselect helpers to select
+#'   elements.
+#' @examples
+#' # Use numeric vector of positions select elements to change:
+#' iris %>% map_at(c(4, 5), is.numeric)
+#'
+#' # Use vector of names to specify which elements to change:
+#' iris %>% map_at("Species", toupper)
+#
+#' @export
+map_at <- function(.x, .at, .f, ...) {
+
+  where <- at_selection(names(.x), .at)
+  sel <- inv_which(.x, where)
+
+  out <- list_along(.x)
+  out[sel]  <- map(.x[sel], .f, ...)
+  out[!sel] <- .x[!sel]
+
+  set_names(out, names(.x))
+}
+
+
+#' @rdname map_if
+#' @description * `map_depth()` allows to apply `.f` to a specific
+#'   depth level of a nested vector.
+#' @param .depth Level of `.x` to map on. Use a negative value to
+#'   count up from the lowest level of the list.
+#'
+#'   * `map_depth(x, 0, fun)` is equivalent to `fun(x)`.
+#'   * `map_depth(x, 1, fun)` is equivalent to `x <- map(x, fun)`
+#'   * `map_depth(x, 2, fun)` is equivalent to `x <- map(x, ~ map(., fun))`
+#' @param .ragged If `TRUE`, will apply to leaves, even if they're not
+#'   at depth `.depth`. If `FALSE`, will throw an error if there are
+#'   no elements at depth `.depth`.
+#' @examples
+#'
+#' # Use `map_depth()` to recursively traverse nested vectors and map
+#' # a function at a certain depth:
+#' x <- list(a = list(foo = 1:2, bar = 3:4), b = list(baz = 5:6))
+#' str(x)
+#' map_depth(x, 2, paste, collapse = "/")
+#'
+#' # Equivalent to:
+#' map(x, map, paste, collapse = "/")
+#' @export
+map_depth <- function(.x, .depth, .f, ..., .ragged = FALSE) {
+  if (!is_integerish(.depth, n = 1, finite = TRUE)) {
+    abort("`.depth` must be a single number")
+  }
+  if (.depth < 0) {
+    .depth <- pluck_depth(.x) + .depth
+  }
+
+  .f <- as_mapper(.f, ...)
+  map_depth_rec(.x, .depth, .f, ..., .ragged = .ragged, .atomic = FALSE)
+}
+
+map_depth_rec <- function(.x,
+                          .depth,
+                          .f,
+                          ...,
+                          .ragged,
+                          .atomic) {
+  if (.depth < 0) {
+    abort("Invalid depth")
+  }
+
+  if (.atomic) {
+    if (!.ragged) {
+      abort("List not deep enough")
+    }
+    return(map(.x, .f, ...))
+  }
+
+  if (.depth == 0) {
+    return(.f(.x, ...))
+  }
+
+  if (.depth == 1) {
+    return(map(.x, .f, ...))
+  }
+
+  # Should this be replaced with a generic way of figuring out atomic
```

Comment:
Can we now use `vec_is_list()`?

Reply:
Yeah, tracking in https://github.com/tidyverse/purrr/issues/920